src/getAvailable.js

```diff
@@ -15,37 +15,38 @@ const getAvailable = async (req, res) => {
     return unauthorized(req, res);
   }
 
-  let datasets;
-  let narratives;
+  let datasetPaths;
+  let narrativePaths;
 
   try {
-    datasets = await source.availableDatasets() || [];
-    narratives = await source.availableNarratives() || [];
+    datasetPaths = await source.availableDatasets() || [];
+    narrativePaths = await source.availableNarratives() || [];
   } catch (err) {
     if (err instanceof ResourceNotFoundError) {
       return res.status(404).send("The requested URL does not exist.");
     }
   }
 
-  if (!datasets || !datasets.length) {
+  if (!datasetPaths || !datasetPaths.length) {
     utils.verbose(`No datasets available for ${source.name}`);
   }
-  if (!narratives || !narratives.length) {
+  if (!narrativePaths || !narrativePaths.length) {
     utils.verbose(`No narratives available for ${source.name}`);
   }
 
-  return res.json({
-    datasets: datasets.map((path) => ({
-      request: joinPartsIntoPrefix({source, prefixParts: [path]}),
-      secondTreeOptions: source.secondTreeOptions(path),
-      buildUrl: source.name === "community"
-        ? `https://github.com/${source.repo}`
-        : null
-    })),
-    narratives: narratives.map((path) => ({
-      request: joinPartsIntoPrefix({source, prefixParts: [path], isNarrative: true})
-    }))
-  });
+  const datasets = [];
+  const narratives = [];
+  for (let i=0; i<datasetPaths.length; i++) {
+    const request = await joinPartsIntoPrefix({source, prefixParts: [datasetPaths[i]]}); // eslint-disable-line no-await-in-loop
+    const secondTreeOptions = source.secondTreeOptions(datasetPaths[i]);
+    const buildUrl = source.name === "community" ? `https://github.com/${source.repo}` : null;
+    datasets.push({request, secondTreeOptions, buildUrl});
+  }
+  for (let i=0; i<narrativePaths.length; i++) {
+    const request = await joinPartsIntoPrefix({source, prefixParts: [narrativePaths[i]], isNarrative: true}); // eslint-disable-line no-await-in-loop
+    narratives.push({request});
+  }
+  return res.json({datasets, narratives});
```

Comment:
Thanks for the review! I'll take a look at implementing this, however my understanding was that something like:
```js
const data = arr.map((item) => ({
   x: (await someAsyncFunction())
})
```
was problematic because `Array.map()` is a synchronous function which results in the `data` array being created with entries containing (keys with values which are) potentially/probably/definately unfulfilled promises. I know one can get around this via
```js
const data = await Promise.all(arr.map(async (item) => ({
   x: (await someAsyncFunction())
})
```
but I preferred loops over this form.

P.S. Why are three clause loops an "anti-pattern"? Can completely understand that people might not like the style, but I've used them a lot (in various languages) and would be interested in learning why they are discouraged.

Reply:
Ah, right. Indeed, the async declaration and `Promise.all` are necessary. I prefer that functional approach to an imperative loop that builds up an array because the structure of the code ends up matching the data structure produced.

Either way, I'm realizing just now that we should think about parallelization of requests here, since that's what's being repeatedly awaited by this code. (More on that in another comment.)

Re: C-style for loops: It's really easy to get the start/stop condition subtly wrong in a C-style for loop, resulting in off-by-one errors (for example). JavaScript, like Python and many other langs, have explicit syntax for iterating over arrays (and many other iterables, like generators) precisely so the intent of the code can be expressed more directly and correctly than index calculations and bounds testing. It's syntax that aims to design away a whole class of bugs and avoid a small form of _accidental complexity_. Like any pattern/anti-pattern, any one instance of it is not a big deal in isolation; it's in aggregate that the gains are made over time. (The same thing that's true of linting, for example.)

Reply:
>  prefer that functional approach to an imperative loop

Changed to the functional approach

Reply:
> Regardless of approach here, I'm realizing just now that we should think about parallelization of requests here, since that's what's being repeatedly awaited by this code. (More on that in another comment.)

My understanding is that each callback here is async because of `joinPartsIntoPrefix` → `[source.]repoNameWithBranch` → `[source.]branch`, the last of which is a single promise set up in the source's constructor. This promise will only run once per-source-instance, no matter how many methods on that source (e.g. `source.repoNameWithBranch`) access it. 

Related to this, is that each different API handler (getAvailable, getDataset etc) instantiates its own source object, which means that the `[source.]branch` promise (fetch) will run multiple times. A similar situation applies to non-community sources.